Add a dismissable admin notice indicating that the Legacy REST API is not compatible with HPOS

Konamiman opened the pull request:
The notice will appear if the orders table is (or has been) selected as the orders data store in the WooCommerce features settings page, and will disappear when that ceases to be true. Once the notice is dismissed it will never appear again.

![image](https://github.com/woocommerce/woocommerce-legacy-rest-api/assets/937723/41451a7e-7df0-4772-8cb5-cf98450f3245)

## How to test

1. Start with the latest trunk of WooCommerce.
2. Make sure that the legacy REST API is disabled (WooCommerce - Settings - Advanced - Legacy API) and that HPOS is disabled (WooCommerce - Settings - Advanced - Features, select posts table as the orders data store).
3. Activate the plugin in the branch of this pull request. If you need a zip file for the plugin you can generate one checking out the branch of this pull request and running `git archive -o woocommerce-legacy-rest-api.zip HEAD`.
4. Verify that no admin notices appear.
5. Enable the legacy REST API and HPOS. Now the notice should appear.
6. Disable the legacy REST API. The notice will disappear.
7. Enable again the legacy REST API and disable HPOS. Again the notice isn't there.
8. Enable HPOS again. The notice is back.
9. Dismiss the notice. Verify that it doesn't appear again no matter what you do with the legacy REST API and HPOS settings.
10. Verify that the legacy REST API continues to function as usual, e.g. query `http://localhost/wc-api/v3/orders`.
11. Switch WooCommerce to the branch of https://github.com/woocommerce/woocommerce/pull/40627 and repeat the tests (note that now you can't disable the legacy REST API, so only test enabling and disabling HPOS).

Note: run this if you want to undo the notice dismissal:

```bash
 wp user meta delete <your user id> "dismissed_legacy_rest_api_is_incompatible_with_hpos_notice"
```

Konamiman committed:
Add a dismissable admin notice indicating that the Legacy REST API
is not compatible with HPOS.

The notice will appear if the orders table is (or has been) selected
as the orders data store in the WooCommerce features settings page,
and will disappear when that ceases to be true.
Once the notice is dismissed it will never appear again.

Konamiman committed:
Update version number, add entry in readme

Konamiman committed:
Small fix that had been incorporated into Woo core

vedanshujain commented on includes/class-wc-legacy-rest-api-plugin.php:
```diff
@@ -58,14 +68,57 @@ public static function on_plugin_activated() {
+     * @returns bool True if the notice has been added, false otherwise.
+     */
+    private static function maybe_add_hpos_incompatibility_admin_notice() {
+        if( 'yes' !== get_option( 'woocommerce_custom_orders_table_enabled' ) || self::user_has_dismissed_admin_notice( 'legacy_rest_api_is_incompatible_with_hpos' ) ) {
```
Should this use the `OrderUtils` check instead of accessing option directly?

Konamiman replied:
Makes sense. I have pushed the change.

vedanshujain reviewed (commented):
Looks good, added a small comment. By the way, not that we have published [our goodbye](https://developer.woocommerce.com/2024/05/14/goodbye-legacy-rest-api/) should we link it using the filter we added in the core PR?

Konamiman committed:
Use the OrderUtil class to check if HPOS is enabled.

Konamiman commented:
> not that we have published [our goodbye](https://developer.woocommerce.com/2024/05/14/goodbye-legacy-rest-api/) should we link it using the filter we added in the core PR?

I think that all the notices and warnings we've added through time in WooCommerce core, which all point to the original announcement, should be enough.
